recipes/serializers.py

```diff
@@ -41,5 +41,37 @@ def create(self, data):
 
         return recipe
 
+    def update(self, instance, data):
+        if "ingredients" in data:
+            new_ingredients = set()
+            for ingredient in data["ingredients"]:
+                lookup_name = ingredient["name"].lower()
+                lookup_name = lookup_name.replace(" ", "_")
+                ingredient, created = Ingredient.objects.get_or_create(name=ingredient["name"], lookup_name=lookup_name)
+                new_ingredients.add(ingredient)
+
+            current_ingredients = set()
+            for ingredient in instance.ingredients.all():
+                current_ingredients.add(ingredient)
+
+            unchanging_ingredients = new_ingredients.intersection(current_ingredients)
+            ingredients_to_add = new_ingredients.difference(unchanging_ingredients)
+            ingredients_to_remove = current_ingredients.difference(unchanging_ingredients)
+
+            for ingredient in ingredients_to_add:
+                instance.ingredients.add(ingredient)
+
+            for ingredient in ingredients_to_remove:
+                instance.ingredients.remove(ingredient)
+
+            data.pop("ingredients", None)
```

Comment:
i am lost :)  I think you can simplify ingredient updating.

Reply:
I know it is kind of over-engineering, but let me explain myself :D. I thought that while updating ingredients there is no need to remove all current ingredients and add incoming ingredients, instead I have found which ingredients need to be added and which ingredients need to be removed. I can switch this code to the delete&add solution though.

Reply:
Still there is a room for improvements.